### PR TITLE
#2632 Refactoring request: improve Editor tools in 'ketcher-react'

### PR DESCRIPTION
**Pull request by kaluginserg**
- in the folder `/packages/ketcher-react/src/script/editor/tool` 
  - create interface AbstractTool with optional methods like: 'mousedown', 'mousemove', 'cancel' ...
  - for each tool in the folder implement the interface
  - for each tool make the strictest modifiers for the private properties and methods
  - remove unused properties
- in the file Editor `/packages/ketcher-react/src/script/editor/Editor.ts`
  - add a new type for the '_tool' property.
  - update related interfaces and properties

### CODE REVIEW

File: packages/ketcher-react/src/script/editor/tool/AbstractTool.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+export interface AbstractTool extends ToolEventHandler {
+  close?(): void
+
+  cancel?(): void
+
+  isSelectionRunning?(): boolean
+
+  isNotActiveTool?: boolean
+}
```
Comment (yuleicul) [inline packages/ketcher-react/src/script/editor/tool/AbstractTool.ts:29]: I'm wondering why `close()`, `isSelectingRunning()`, `isNotActiveTool()` should be in AbstractTool. I thought that not every tool needed them, right?

File: packages/ketcher-react/src/script/editor/tool/AbstractTool.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  mouseup?(event: Event): void
+
+  mouseleave?(event: Event): void
+
+  mouseLeaveClientArea?(event: Event): void
+
+  mouseover?(event: Event): void
+}
+
+export interface AbstractTool extends ToolEventHandler {
```
Comment (yuleicul) [inline packages/ketcher-react/src/script/editor/tool/AbstractTool.ts:21]: I thought TypeScript had a keyword [abstract](https://www.typescriptlang.org/docs/handbook/2/classes.html#abstract-classes-and-members) which was specific for abstract class. 

I'm not sure if we can:
```suggestion
abstract class ToolEventHandler {}
abstract class Tool extends ToolEventHandler {}
```

I thought the advantage of `interface` was there would be no extra JavaScript output, but one good thing about `abstract` was if we forget to implement the base class’s abstract members, we’ll get an error.

What's your take on this?

File: packages/ketcher-react/src/script/editor/tool/AbstractTool.ts (hunk 1 above)
Comment (kaluginserg) [reply packages/ketcher-react/src/script/editor/tool/AbstractTool.ts:29]: actually, it `isSelectingRunning()` and `isNotActiveTool` is something that is needed for every tool, as it is an interface used from the outside to control abstract tool.

Example `Editor.ts:633`:
```
  const EditorTool = editor.tool()
  ....
    clientArea.contains(event.target) || EditorTool.isSelectionRunning?.(),
  ]
```

Example `Editor.ts:234`:
```
if (!tool || tool.isNotActiveTool) {
```


Even I think, we should mark those methods as required and implement them in each tool. Consistency leads to simplicity.

Not sure about `.close()` - can't find the case... Looks like it is my mistake.

@yuleicul what do you think about it?

File: packages/ketcher-react/src/script/editor/tool/AbstractTool.ts (hunk 2 above)
Comment (kaluginserg) [reply packages/ketcher-react/src/script/editor/tool/AbstractTool.ts:21]: I doubt using the base class as it often tends to use real inheritance with class extends and common logic, which is often tough to manage.
I agree that the 'Abstract' word could lead to misunderstanding. 
So my proposal is to rename `AbstractTool` to `Tool` and use it as is as a common interface.

@yuleicul let me know if you have other ideas

File: packages/ketcher-react/src/script/editor/tool/AbstractTool.ts (hunk 1 above)
Comment (yuleicul) [reply packages/ketcher-react/src/script/editor/tool/AbstractTool.ts]: Thanks for your examples, make sense

**Review by yuleicul — APPROVED** [review]

**Pull request merged**